### ISSUE TITLE
How to fully process long files and ignore silences?

### ISSUE THREAD

**Issue by mrscarvalho** [Azure-Samples/cognitive-services-speech-sdk#1335]
I’m testing some speech-to-text SDKs, and Azure is providing the best results so far (our content is difficult because usually has background music, fast paced speech or even sung text).

But I’m not being able to process the whole files, that are provided in audio files between 30 seconds up to 3 minutes.
 
While recognize_once is limited to 15”, ContinuousRecognition is controlled by SILENCES, what is very annoying when we want to simply process whole files. 

First it returns empty text if there is silence in the beginning. Or interrupts in the middle before the end if finds silence. I found some non-official solutions (sending parameters) to make it accept longer silences, but now it does not stop at the end of file, remaining “listening” for a long time. Another problem is that "non transcriptable" speech or music are considered silence as well (should be simply ignored instead), so simply removing real silences does not help (I did using FFMPEG).

Is there some other option, just like the other SDKs do, to simply process the whole file until the end? Maybe extend the 15” time limit of recognize_once?

An option would be to be able to provide for ContinuousRecognition the exact duration of the file, and it goes until this point ignoring any silence. Is there this possibility? I cannot do that simply controlling delayed time because I don’t know how much of the file was sent and processed.

Another point is that we do not need real time transcription, usually file processing is much faster than real time (ex: 30 seconds audio may be processed in 10 or 15”).

Thanks in advance!

**Comment by rhurey** [comment]
First, thanks for using the SDK and letting us know where you're having issues.

ContinuousRecognition should be the correct API to use to recognize a long file, regardless of the amount of silence in the file.

Is the SpeechRecognizer object raising a Canceled event during the file processing? If it is can you share the Reason and ErrorDetails properties from that event?

What I'd expect to see raised as events from the SDK would be:
(Names based on C#, other languages are similar)

1. SessionStarted (Indicates the session started)
2. SpeechStartDetected (When the first speech is found in each 'turn')
3. Recognizing (Intermediate results)
4. Recognized (Final results for a single utterance)
5. SpeechEndDetected (Indicates speech has stopped in a single 'turn')
6. Canceled (Reason == EndOfStream at the end of the file)
7. SessionStopped (Indicates Recognition is finished)

For 2 through 5, there can be multiple iterations based on the amount of silence in the audio stream, the length of the audio (there's a maximum service connection length) and any network errors that cause the SDK to reconnect to the service. None of those reasons should stop recognition.

The reason for having interim events when there's silence in the audio stream is to help the developer who is working on an interactive user scenario know that the person has stopped talking in a deterministic way.

How are you supplying audio to the SDK? You mentioned it remains "listening" and I'm interested in why that is.

**Comment by mrscarvalho** [comment]
Hello @rhurey , thank you for your reply and detailed information about the events, that helped a lot to understand the way the continuous recognition works. Seeing the big picture of the continuous process I could see the great work your team has done providing almost real time continuous recognition, and providing many options for developers to work.

After several tests during this week I realized that a mistake I did myself related to a debug flag that I introduced caused the non-stop recognition (it prevented the cancel event to be connected to the stop function, so it was stopping only at the timeout in the waiting loop). After fixing it the recognition is working properly and stopping at the end of the file. About the events, I could see that it is enough to monitor “Recognized” event to get pieces of text without repetition, and that “SpeechEndDetected” always rise a “Canceled” event that can trigger the stop command. 

About silence, doing more tests I also realized that Continuous Recognition was not holding due silence, although I’m using a FFMPEG pre-processing to generate the WAV file that removes any silence larger than 1 second (also shortens the whole file). The issue with silence happened previously, when I was testing “recognize_once_async” (sorry the confusion), but was solved using initialSilenceTimeoutMs and endSilenceTimeoutMs parameters at the endpoint URL (but the 15s limitation will force the usage of continuous detection anyway). In the last tests I had no unexpected Canceled events, but I will keep testing to check if all files are being fully processed.

Time processing is taking about 65% of the audio duration to be processed (I will do performance comparison and heavier tests after setting up all APIs). Currently, our priority is related to transcription quality, as our contents is difficult and although preliminar good results comparing to previous solution used, there is still a reasonable amount of mistakes (it is mostly TV and radio contents with background music, noises and often fast paced speech, sometimes even sung).

Sorry for the confusion and bugs, unfortunately there is no permanent listening (I’m sending simple WAV files at once, not streams). :-)
